src/cpp/logger.cpp

```diff
@@ -0,0 +1,252 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <cstdlib>
+#include <string>
+#include <ctime>
+#include <iostream>
+#include <fstream>
+#include <iomanip>
+
+#include "utility.h"
+#include "logger.h"
+#include "srexception.h"
+
+using namespace SmartRedis;
+
+/*!
+*   \brief Rename the current client
+*   \param _client_id new ID to use for the current client
+*/
+void Logger::rename_client(const std::string& client_id)
+{
+    _client_id = client_id;
+}
+
+
+/*!
+*   \brief Set up logging for the current client
+*   \param _client_id ID to use for the current client
+*/
+void Logger::configure_logging(const std::string& client_id)
+{
+    // If we're already initialized, they can set up a client ID
+    // Useful if they call a Dataset API point before setting up
+    // a Client object
+    _client_id = client_id;
+    if (_initialized) {
+        return;
+    }
+    _initialized = true;
+
+    // Get the logfile
+    get_config_string(_logfile, "SR_LOG_FILE", "", true);
+    std::string requestedLogfile(_logfile);
+    bool missingLogFile = _logfile.length() == 0;
+
+    // Make sure it can be written to
+    bool badLogFile = false;
+    if (_logfile.length() > 0) {
+        std::ofstream logstream;
+        logstream.open(_logfile, std::ios_base::app);
+        badLogFile = !logstream.good();
+
+        // Switch to STDOUT if we can't write to the specified file
+        if (badLogFile)
+            _logfile = "";
+    }
+
+    // Get the logging level
+    std::string level;
+    get_config_string(level, "SR_LOG_LEVEL", "", true);
+    bool missingLogLevel = level.length() == 0;
+    bool badLogLevel = false;
+    if (level.length() > 0) {
+        if (level.compare("QUIET") == 0)
+            _log_level = LLQuiet;
+        else if (level.compare("INFO") == 0)
+            _log_level = LLInfo;
+        else if (level.compare("DEBUG") == 0)
+            _log_level = LLDebug;
+        else if (level.compare("DEVELOPER") == 0)
+            _log_level = LLDeveloper;
+        else {
+            // Don't recognize the requested level; use default
+            badLogLevel = true;
+            _log_level = LLInfo;
+        }
+    }
+    else {
+        _log_level = LLInfo;
+    }
+
+    // Now that everything is configured, issue warning and
+    // error messages
+    if (missingLogFile) {
```

Comment:
Yes! It gives us a chance to fully establish logging and log some errors the way the customer wants them logged instead of in our default method. This in turn will improve the chances of the errors being noticed.